compiler/src/dotty/tools/dotc/core/NullOpsDecorator.scala

```diff
@@ -0,0 +1,144 @@
+package dotty.tools.dotc.core
+
+import dotty.tools.dotc.core.Contexts.Context
+import dotty.tools.dotc.core.Symbols.defn
+import dotty.tools.dotc.core.Types.{AndType, ClassInfo, ConstantType, OrType, Type, TypeBounds, TypeMap, TypeProxy}
+
+/** Defines operations on nullable types. */
+object NullOpsDecorator {
+
+  implicit class NullOps(val self: Type) {
+    /** Is this type a reference to `Null`, possibly after aliasing? */
+    def isNullType(implicit ctx: Context): Boolean = self.isRef(defn.NullClass)
+
+    /** Is this type exactly `JavaNull` (no vars, aliases, refinements etc allowed)? */
+    def isJavaNullType(implicit ctx: Context): Boolean = {
+      assert(ctx.explicitNulls)
+      // We can't do `self == defn.JavaNull` because when trees are unpickled new references
+      // to `JavaNull` could be created that are different from `defn.JavaNull`.
+      // Instead, we compare the symbol.
+      self.isDirectRef(defn.JavaNullAlias)
+    }
+
+    /** Normalizes unions so that all `Null`s (or aliases to `Null`) appear to the right of
+     *  all other types.
+     *  e.g. `Null | (T1 | Null) | T2` => `T1 | T2 | Null`
+     *  e.g. `JavaNull | (T1 | Null) | Null` => `T1 | JavaNull`
+     *
+     *  Let `self` denote the current type:
+     *  1. If `self` is not a union, then the result is not a union and equal to `self`.
+     *  2. If `self` is a union then
+     *    2.1 If `self` does not contain `Null` as part of the union, then the result is `self`.
+     *    2.2 If `self` contains `Null` (resp `JavaNull`) as part of the union, let `self2` denote
+     *      the same type as `self`, but where all instances of `Null` (`JavaNull`) in the union
+     *      have been removed. Then the result is `self2 | Null` (`self2 | JavaNull`).
+     */
+     def normNullableUnion(implicit ctx: Context): Type = {
+      var isUnion = false
+      var hasNull = false
+      var hasJavaNull = false
+      def strip(tp: Type): Type = tp match {
+        case tp @ OrType(lhs, rhs) =>
+          isUnion = true
+          val llhs = strip(lhs)
+          val rrhs = strip(rhs)
+          if (rrhs.isNullType) llhs
+          else if (llhs.isNullType) rrhs
+          else tp.derivedOrType(llhs, rrhs)
+        case _ =>
```

Comment:
Consider normalizing also in intersections, the way it is done in `OrNull.stripNull`.

Reply:
Normalizing intersections is added.